### PR TITLE
Introduce edgestitch

### PR DESCRIPTION
**Pull request by xjunior**
Edgestitch is a rethought approach to the `db:stitch` tasks extracted from nitro. While in nitro, the process was to form the structure.sql from the structure-self.sql of components that the target component depends on, using cobra to discover this dependency tree. 

This new approach finds all loaded Engines, and puts together their structure-self.sql to form a new structure.sql, allowing external engines also to provide and own tables. The end result is the same, without the CobraCommander overhead.

### CODE REVIEW

File: packages/edgestitch/Gemfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+source "https://rubygems.org"
+
+gemspec
+
+rails_version = ENV.fetch("RAILS_VERSION", ">= 5.2.8.1")
```
Comment (that-jill) [inline packages/edgestitch/Gemfile:7]: are we supporting rails 5.2?

Comment (xjunior) [reply packages/edgestitch/Gemfile:7]:
```suggestion
rails_version = ENV.fetch("RAILS_VERSION", ">= 6.0.6")
```

File: packages/edgestitch/lib/edgestitch/version.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+module Edgestitch
+  VERSION = "1.0.0"
```
Comment (that-jill) [inline packages/edgestitch/lib/edgestitch/version.rb:4]: are we ready to call this "good" and not "experimental" ?

File: packages/edgestitch/spec/dummy/config/initializers/application_controller_renderer.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+# Be sure to restart your server when you modify this file.
+
+# ActiveSupport::Reloader.to_prepare do
```
Comment (that-jill) [inline packages/edgestitch/spec/dummy/config/initializers/application_controller_renderer.rb:4]: Do we need these files?

**Review by that-jill — COMMENTED** [review]
lots of questions, overall I like it a lot

File: packages/edgestitch/docs/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+## Usage
+
+Edgestitch will enhance the default rails tasks, so using nothing special has to be done in order to use it. Once edgestitch is correctly installed things should Just Work™️ as explained in the rails manual, but it will be generating `structure-self.sql` along with `structure.sql` files. **Important**: It's recommended that `structure.sql` files are added to `.gitignore`.
+
+# How does it work
+
+Edgestitch works based on table and migrations ownership. Based on these ownerships, Edgestitch can export a `structure-self.sql` file defining the DDL owned by a specific engine.
+
+The stitching process then takes all loaded engines (assuming the dependency between engines is defined correctly) and assembles a structure.sql file.
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/docs/README.md:73]: The problem would be exactly the same.

File: packages/edgestitch/docs/README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+## Table and Migration Ownerships
+
+A model is owned by an engine when it inherits from the Engine's ApplicationModel, and thus the table is owned by that engine. A migration is owned by the engine if it is defined within its `db/migrate` directory.
+
+## Extra tables
+
+Sometimes an external dependency brings in extra tables that are not defined in any `structure-self.sql`. To be part of the ecosystem, the new tables should be owned by the engine adding them. That can be done by adding these tables to `<engine>/db/extra_tables`. It's a simple text file with a list of table names that do not inherit from `<Engine>::AplicationModel`, but are owned by that engine and should be included in its structure-self.sql.
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/docs/README.md:81]:
```suggestion
When an external dependency brings in extra tables (i.e. acts_as_taggable) that are not defined in any `structure-self.sql`. To be part of the ecosystem, the new tables should be owned by the engine adding them. That can be done by adding these tables to `<engine>/db/extra_tables`. It's a simple text file with a list of table names that do not inherit from `<Engine>::AplicationModel`, but are owned by that engine and should be included in its structure-self.sql.
```

What about this?

File: packages/edgestitch/docs/README.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+## External Gems
+
+An external gem can also define a `structure-self.sql` file to be adapted in this ecosystem. That can be done using the same approach specified in Installation / Engines.
+
+## Development
+
+After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.
+
+To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/docs/README.md:91]:
```suggestion
To install this gem onto your local machine, run `bundle exec rake install`.
```

File: packages/edgestitch/lib/edgestitch/exporter.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    #
+    def self.export(engine, dump)
+      new(engine).export(dump)
+    end
+
+    attr_reader :engine
+
+    def initialize(engine)
+      @engine = engine
+      @database_directory_path = engine.root.join("db")
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/exporter.rb:23]: We can do that, can I save this for later?

File: packages/edgestitch/lib/edgestitch/exporter.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    def models
+      @models ||= application_record&.descendants || []
+    end
+
+    def application_record
+      @application_record ||= engine.railtie_namespace&.const_get(:ApplicationRecord, false)
+    rescue LoadError, NameError
+      nil
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/exporter.rb:69]: What do you mean?

File: packages/edgestitch/lib/edgestitch/version.rb (hunk 2 above)
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/version.rb:4]: I think so… the test this went through was basically to be successful with the 150 nitro components + produce the same SQL as the previous implementation.

File: packages/edgestitch/spec/dummy/config/initializers/application_controller_renderer.rb (hunk 3 above)
Comment (xjunior) [reply packages/edgestitch/spec/dummy/config/initializers/application_controller_renderer.rb:4]: No… and I hate that rails generate them for us.

File: packages/edgestitch/spec/edgestitch/tasks_spec.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RAILS_62_ABOVE = Rails.version >= "6.2"
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/spec/edgestitch/tasks_spec.rb:5]: Well, we support pretty much everything >= 6.0. What exactly are you suggesting?

File: packages/edgestitch/lib/edgestitch/exporter.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/exporter.rb:23]: Done

File: packages/edgestitch/docs/README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+## Contributing
+
+Bug reports and pull requests are welcome on GitHub at https://github.com/powerhome/power-tools.
+
+## License
+
+The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).
```
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/docs/README.md:99]: <3

File: packages/edgestitch/lib/edgestitch/exporter.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/exporter.rb:71]: `LoadError` could happen when a file that should define a constant doesn't, and `autoload` blows it up. `NameError` will raise when `ApplicationRecord` is undefined.

File: packages/edgestitch/docs/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (that-jill) [reply packages/edgestitch/docs/README.md]: This clarifies so much! It also brings up more questions of how we need to educate developers on this gotcha. Will this problem always be caught in CI or is it possible for the issue to get merged and be there for days/weeks before anyone notices?

File: packages/edgestitch/lib/edgestitch/exporter.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (that-jill) [reply packages/edgestitch/lib/edgestitch/exporter.rb:71]: My question here is, I guess, why are we swallowing these errors and how do we surface issues to the user? What about if something random that is neither of those breaks? Does it correctly point to edgestitch or do we have to dig through a weird stack to find it?

[the remark this answers is not in the document]
Comment (xjunior) [reply packages/edgestitch/lib/edgestitch/exporter.rb:71]: @that-jill they will only happen during an export task (within Edgestitch), and they're pretty much expected as not all engines will necessarily define an ApplicationRecord. The (expected) behavior is that the component won't export any tables.

File: packages/edgestitch/spec/edgestitch/tasks_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (that-jill) [reply packages/edgestitch/spec/edgestitch/tasks_spec.rb:5]: let's talk about this PR IRL. seems more efficient that way.

**Review by that-jill — APPROVED** [review]
<3

**Pull request merged**